Feat/build wheels improvement

jakub-kocka opened the pull request:
- Added `include_list` and `exclude_list` for additional Python packages or temporary workaround
    - **clear prints of changes to assembled requirements**
- Edited `README.md`
    - refactor 
    - explanation of usage `exclude_list`/`include_list`
- Changed workflow name and extended it for every platform
    - One workflow file, is used for all platforms instead of separated files for platforms
- Added wheels build
- Converted upload of wheels to Python script
- Added Python version option into yaml file

[workflow run](https://github.com/espressif/idf-python-wheels/actions/runs/7142131695/job/19450764972?pr=21) under "Build wheels for IDF" job step is clear output of what requirements are downloaded and what changes have been done after applying requirements from `exclude_list` and `include_list`
    - MacOS M1 architecture is missing because of paid runner to not waste GitHub Actions spending limit ([M1 workflow](https://github.com/espressif/idf-python-wheels/actions/runs/6971282866/job/18970992725))
    
   Following steps:
   - Make ad-hoc work (manual specification of requirement/s to be build and upload wheels for)
   - Clean up the repository
   - Migrate and change default branch to `master` ?

fhrbata commented on build_wheels.py:
```diff
@@ -172,28 +177,198 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+                        ver_txt += f'{_change_specifier_logic(ver_str)},'
+                    else:
+                        ver_txt += f'{ver_str},'
+                req_txt += ver_txt[:-1]
```
Am I understand it correctly that you are adding ',' and here you are removing it. What's the logic behind this? Wouldn't it be just easier to keep the `version` as a plain string and do a split on ','. With this you can also remove the list test condition and handle all versions in the same way.

Thank you

jakub-kocka replied:
Hi, @fhrbata thank you for review.
Yes exactly, there is hard-coded  ',' in for loop and this char is deleted after last element to not mess up the syntax for pip defined by [PEP440](https://peps.python.org/pep-0440/).

It is not bad idea to have it in plain text but I think the usage of `YAML` file would be almost pointless.

In fact `exclude_list` and `include_list` could be done with `txt` files with valid syntax. For example: 
    
    "dbus-python<3,>=2;sys_platform=='win32'"

This example is valid Requirement from `packaging` package and can be directly used with pip, which is handling build of wheel.

But to made it easier and quicker for developer to add into this lists and not think about opposite logic when exclude is wanted or if the syntax is right the YAML files are used with (I hope) easy and more likely error-free approach.

However you are right it does look a bit weird, I will try to improve this. Also I have found a mistake in implementation of `yaml_to_requirement` function so it will need some changes, thank you.

fhrbata replied:
@jakub-kocka , thank you for additional information. What I had in mind is something like this
```
            ver_list = [package['version']] if type(package['version']) is not list else package['version']
            ver_list = [_change_specifier_logic(v) if exclude or v for v in ver_list]
            req_txt += ','.join(ver_list)
```
not tested, but I think it should do the same thing. Just a though, feel free to ignore.

jakub-kocka replied:
Sorry, now I get what you have originally meant. Definitely better solution, thanks.

I misunderstood that you meant the `YAML` format of version should be something like `version: '<3,>=2'`.

fhrbata replied:
Actually both, sorry if I worded it wrong. After your explanation I realized that using `version: '<3,>=2'` is probably not the best idea, but I personally wouldn't mind it :). Other thing was the different handling of str and list. Anyway these are just nitpicks, so I resolved this. Thank you!

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+        # TXT file from all downloaded requirements and constraints files
+        # useful for debugging or to see the comments for requirements
+        with open('requirements.txt', 'w') as f:
+            for line in requirements_txt:
```
(optional) I think this should do the same and it is shorter and more Pythonic:
```python
f.write('\n'.join(requirements_txt))
```

dobairoland commented on .github/workflows/build-wheels-platforms.yml:
```diff
@@ -0,0 +1,179 @@
+        include:
+          - os: linux-armv7-self-hosted
+            CONTAINER: python:3.8-bullseye
+          - os: BrnoRPI-GH004
```
Could this be something like `linux-rpi-self-hosted`? We try to hide our server locations from our repos.

jakub-kocka replied:
Yes, sure. 
But I think we will need to add new tag for this runner, is it right @tomassebestik? Because `ARM64` tag is used for Linux and MacOS runners, the Linux machine is higher in the list but I am not sure which mechanism is used to select the correct runner.

tomassebestik replied:
@jakub-kocka Normally we would use combination `self-hosted`+`arm64`, like for example [here](https://github.com/espressif/pytest-embedded/blob/main/.github/workflows/test-python.yml#L14-L16).

I'm not sure this would be easy in your workflow .... I have added new tag `linux-arm64-self-hosted` to BrnoRPI-GH004 runner, you can use that.

<img src="https://github.com/espressif/idf-python-wheels/assets/33937168/0aef1dbd-2d9b-4778-8ffa-d53ca4b5a184" width=500px>

jakub-kocka replied:
Thank you.

dobairoland commented on exclude_list.yaml:
```diff
@@ -0,0 +1,11 @@
+# List of Python packages to exclude from automatically assembled requirements
+#"From assembled <main requirements> exclude <package_name> with <version> on <platform>".
+
+# exclude_list template
```
How can we specify Python version restrictions, e.g. package cannot be build with specific Python versions?

jakub-kocka replied:
Since we are building only for Python 3.8 I think it will basically have no effect to do so (to build or not to build should be sufficient).
But for some wheels which are Python version specific maybe we should build the wheels for other Python versions and specify the version in exclude/include lists will be needed.

Maybe this is the topic for further discussion in person?

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+    excluded_requirements = exclude_from_requirements(requirements, exclude_list)
+
+    include_list = yaml_to_requirement('include_list.yaml')
+    print(Fore.BLUE + '---------- ADDITIONAL REQUIREMENTS ----------' + Style.RESET_ALL)
```
(optional) We can avoid all of these string concatenations:

```python
print(str1, str2, str3)
```

Also, probably we could add wrappers like `print_blue()`, or similar.

dobairoland replied:
Also please check all prints. Several colors are used but not for the warning in line 43.

tomassebestik replied:
It might be worth introducing a helper function if it is frequently used...

See in this project: 
- https://github.com/espressif/conventional-precommit-linter/blob/master/conventional_precommit_linter/helpers.py#L8-L9
- https://github.com/espressif/conventional-precommit-linter/blob/master/conventional_precommit_linter/hook.py#L160

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+        - e.g. "<3" will be ">3"
+    """
+    new_spec = specifier
+    replacements = {'<': '>',
```
You don't use lookup so hashing is an unused overhead. Better use a data types which is better for iteration.

```python
(('<', '>',), ('>', '<'))
```

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+                    requirements_set.add(Requirement(requirement_old))
+
+
+        requirements_set.add(Requirement(requirement_txt))
```
You do a lot of string concatenation in this function. You could just collect the strings into a list and join them at the end. 
```python
str_list.append('str1')
str_list.append('str2')
...
requirements_set.add(Requirement(''.join(str_list)))
```
Appending to string requires to allocate a new array, copy both strings there and de-allocate the original strings. Doing this iteratively is a big overhead. Python cannot optimize it.

The above will do concatenation only once (it can count the right array size in advance to fit there all the strings). Python can optimize it and could improve with future versions. Makes this sense to you?

jakub-kocka replied:
Yes, it makes. Thank you for explanation.

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+        if non_classic_requirement and non_classic_requirement[0].replace('--only-binary ', '') in requirement.name:
+            only_bin = non_classic_requirement[0].replace('--only-binary', '').strip()
+            out = subprocess.run(
+                ['python', '-m', 'pip', 'wheel', f'{requirement}',
```
Would be safer to use `sys.executable` instead of `python`.

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+    dir = f'{os.path.curdir}{(os.sep)}downloaded_wheels'
+    for requirement in requirements:
+        # --only-binary requirement wheel build
+        if non_classic_requirement and non_classic_requirement[0].replace('--only-binary ', '') in requirement.name:
```
Na name non_classic suggest something generic. You've mentioned only-binary as an example. But here `--only-binary` is hardcoded three times. Perhaps we need to make this general (compatible with other similar cases), even only now only-binary will be defined there.

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+    print(Fore.RED + f'Failed {failed_wheels} wheels' + Style.RESET_ALL)
 
-    print_requirements(requirements_set)
+    assert failed_wheels == 0
```
(optional) You could use this to generate an exit code. It would be more user-friendly then throwing an AssertionError.

tomassebestik replied:
To proper handle exit code from the script ... FYI https://github.com/espressif/conventional-precommit-linter/blob/master/conventional_precommit_linter/hook.py#L223-L224

dobairoland commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+                    if not printed:
+                        print(Fore.RED + f'-- {requirement}')
+                        printed = True
+                    print(Fore.GREEN + f'++ {new_requirement}')
```
This looks pretty cool in the output. :-)

dobairoland reviewed (changes requested):
@jakub-kocka Thank you very much! Looks very good, I've left only a couple of minor suggestions.

I've checked the output as well for a couple of known corner examples. All good!

tomassebestik commented on Upload-Wheels.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-cd download || exit
+cd downloaded_wheels || exit
```
Maybe `exit 1` here ... (avoid plain exit).

This script as written now is kind of unreadable and also without handling possible errors.
As minimum, I would suggest check this script with `shellcheck`.
***
Maybe it could be rewritten to make it more readable and robust:
```bash
#!/usr/bin/env bash

usage() {
  echo "Usage: $0 <s3-bucket-name>"
  echo "This script uploads .whl files from the downloaded_wheels directory to the specified S3 bucket."
  exit 1
}

# Validate input
if [ -z "$1" ]; then
  echo "Error: S3 bucket name not provided."
  usage
fi

# Check if AWS CLI installed
if ! command -v aws &> /dev/null; then
  echo "Error: AWS CLI is not installed."
  exit 1
fi

# Change to downloaded_wheels directory
cd downloaded_wheels || { echo "Error: downloaded_wheels directory not found."; exit 1; }

# Check if there are any .whl files
if [ -z "$(ls *.whl 2>/dev/null)" ]; then
  echo "No .whl files found in downloaded_wheels directory."
  exit 1
fi

# Loop over .whl files
for wfile in *.whl; do
  # Process file name
  file_name="${wfile%%-*}"
  file_name="${file_name,,}"  # Convert to lowercase
  file_name="${file_name//_/-}"  # Replace underscores with hyphens

  # Skip esptool for specific OS
  case "$(uname -sr)" in
    CYGWIN*|MINGW*|MINGW32*|MSYS*)
      if [[ "$file_name" == "esptool" ]]; then
        continue
      fi
      ;;
    *)
      ;;
  esac

  # Upload to S3
  if ! aws s3 cp --acl=public-read --no-progress "$wfile" "s3://$1/pypi/$file_name/$wfile"; then
    echo "Error uploading $wfile to S3."
    exit 1
  fi
done

echo "All good ... all files uploaded successfully!"

```

(In case of time pressure, you can ignore this comment for now - just a suggestion.)

tomassebestik commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+                    requirement_old += f"; sys_platform != '{package['platform']}'"
+                    requirements_set.add(Requirement(requirement_old))
+
+            if isinstance(package['platform'], list):
```
`else:` ?

tomassebestik commented on build_wheels.py:
```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+
+    for requirement in assembled_requirements:
+        printed = False
+        for req in exclude_list:
```
This function is very hard to read because of the nesting level ... it might be worth rethinking this design (split the logic into smaller blocks).

In addition, using `requirement` for the outer loop and a look-alike `req` for the inner loop is extra confusing ... a longer more descriptive variable name would be better here.

tomassebestik commented:
Nice improvements, @jakub-kocka! 

I left a few comments about style and readability - feel free to ignore them or address them later if your goal is to merge this as soon as possible.

jakub-kocka commented:
Thank you for your reviews @dobairoland, @tomassebestik I have implemented some of your suggested changes.

And as @fhrbata suggested I have also rewritten uploading of wheels into Python.

According to this I have a question about the usage of uploaded wheels:
Is it used by ESP-IDF or any tool already? Because there are few things I don't understand correctly and I thing it should be improved as well
- Why are we transforming package name to lowercase and changing '_' to '-'? (when it is not actual name of the package)

>  file_name=$(echo "${wfile%%-*}" | tr '[:upper:]' '[:lower:]' | tr '_' '-')

- Is the index formatting strictly needed to be like this or should the page look a bit better? (for example there should be user friendly graphics and JSON file to be used when tools will need the wheels)

- Also it looks like there are some leftovers of badly used logic of `Upload-Wheels.sh` that make no sense and it should be cleaned up in my opinion.

For example `dbus` directory which is empty and should be `dbus-python` I suppose.

[Workflow](https://github.com/espressif/idf-python-wheels/actions/runs/7005203747) after changes.

jakub-kocka commented:
I have added `python` version marker option to yaml file, usage is the same as for the `platform`. It is mentioned in README.md. [Workflow with changes](https://github.com/espressif/idf-python-wheels/actions/runs/7248586503/job/19746445743?pr=21).

Also because the `yaml_to_requirement()` function became a bit complicated I have added tests for this and test_yaml file with all possible combinations.

Added a pre-commit hook to check yaml files syntax to prevent future errors when adding into include/exclude lists.

Added `get_python_dependent_wheels()` function which finds Python version specific wheels and saves it into txt file, this file is after intended to be called from sub-workflow and build Python wheels for other Python versions. This approach should be less time and resource consuming than running the whole workflow again which is not really needed (I have attached simple diagram for better understanding of intended feature)
![wheels_workflow_diag](https://github.com/espressif/idf-python-wheels/assets/46049162/a2a542ab-0b60-4d48-94c2-4b4c9b1024b3)


I need to get this PR merged to be able to call local workflow  `build-wheels-python-dependent.yml` so the described feature is not working yet.

Can you please take a look @dobairoland, @tomassebestik? So I would be able to continue on this and other steps specified in PR description/task.

dobairoland commented on upload_wheels.py:
```diff
@@ -0,0 +1,36 @@
+    wheel_name = wheel_name.lower()
+    wheel_name = wheel_name.replace('_', '-')
+
+    if sys.platform in 'CYGWIN,MINGW,MINGW32,MSYS' and wheel_name == 'esptool':
```
It would be great to avoid these kinds of hacks in the new version. This could be specified with the exclude list.

jakub-kocka replied:
You are right, it is not necessary anymore, thank you.

dobairoland commented on .github/workflows/build-wheels-platforms.yml:
```diff
@@ -0,0 +1,191 @@
+            arch -arm64 python -m pip install 'cython<3'
+            arch -arm64 python -m pip install 'cffi'
+            arch -arm64 python -m pip install 'gevent'
+            arch -arm64 python -m pip install 'gdbgui==0.13.2.0' --no-build-isolation
```
It would be great if don't have to hard code these here. Probably `--no-build-isolation` is the hardest part.

dobairoland replied:
Could the build dependencies go into `build_requirements.txt`?

jakub-kocka replied:
I am not sure if it will work, because it is MacOS M1 specific issue and `build_requirements.txt` is common for all platforms, I will try it.

jakub-kocka replied:
OK, I have tried this ([workflow1](https://github.com/espressif/idf-python-wheels/actions/runs/7272987742), [workflow2](https://github.com/espressif/idf-python-wheels/actions/runs/7273106410)) with no success.

And according to [documentation even the newest pip doesn't support `--no-build-isolation`](https://pip.pypa.io/en/latest/reference/requirements-file-format/#supported-options) so for now I would keep this solution with comment and as soon as it will be possible to change it we should do it than.

dobairoland replied:
Please try to address this in a follow-up PR because one of the main reasons for this refactoring is to avoid this kind of hacks and have these defined in configuration files.

dobairoland commented on build_wheels_from_file.py:
```diff
@@ -0,0 +1,29 @@
+import subprocess
```
Could you please add copyright headers for source files (not just this one)?

dobairoland commented on upload_wheels.py:
```diff
@@ -0,0 +1,36 @@
+wheel_files = os.listdir(WHEELS_DIR)
+
+for wheel in wheel_files:
+    pattern = re.compile(r'([^ -]*)-(\d+)')
```
Is this `[^ -]` isn't (almost) the same as as `\w`?

jakub-kocka replied:
You are right, because of not including only spaces or hyphens  `\w` does the same. Thank you.

jakub-kocka commented on .github/workflows/build-wheels-platforms.yml:
```diff
@@ -0,0 +1,191 @@
+name: platforms-dispatches
+
+on:
+  # TODO schedule on final merge
```
I am not sure if we want to suspend the "old wheels build"?

Because if we do the schedule, both builds will run together. And now for the "new wheels build" (this PR) there is only Python 3.8 but much more wheels and for the "old wheels build" there are more Python versions but only some wheels.

jakub-kocka commented:
@dobairoland I have implemented your suggested changes, [wokflow run](https://github.com/espressif/idf-python-wheels/actions/runs/7273144357/job/19817218607?pr=21).

dobairoland commented:
@tomassebestik @fhrbata I'm merging this so Jakub could continue his work. If you will have comments regarding this later (after you are back to work) then he will address those in a consequent PR.
